### PR TITLE
GCI: FIX 3 for PR

### PR DESCRIPTION
**Pull request by felixfavour**
Fixed all issues mentioned in the last Pull request.

### CODE REVIEW

File: dataservice.js [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+const combinedProductMap = new Map();
+const productMap = new Map();
+const categoryMap = new Map();
+
+// function to load products JSON and convert it to a map.
+let getProducts = () => {
+    const productsJson = require('./my-app-react/src/data/products.json');
+    if (productMap.size === 0) {
+        for (let product of productsJson.products) {
```
Comment (f4ww4z) [inline dataservice.js]: Can't this be const?
```suggestion
        for (const product of productsJson.products) {
```

File: dataservice.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    }
+};
+
+// Function to combine both products and categories in a collectable and renderable format
+let combineProductsWithCategories = () => {
+    getProducts();
+    getCategories();
+
+    for (let product of productMap.products) {
+        for(let category of categoryMap.categories) {
```
Comment (f4ww4z) [inline dataservice.js]:
```suggestion
        for(const category of categoryMap.categories) {
```

File: server.js [hunk 3]
```diff
@@ -1,49 +1,41 @@
...
-const app = express()
+const express = require('express');
+const app = express();
 
 app.get('/info', (req, res) => {
-    info = [
-        {"serverName": "localhost5000"},
+    let info = [
+        {"serverName": "localhost5001"},
         {"sampleName": "chinemerem-react"},
```
Comment (f4ww4z) [inline server.js]: Did you address my last comment?
```suggestion
        {"serverName": "chinemerem-react"},
```
Remove `sampleName`.

File: server.js [hunk 4]
```diff
@@ -1,49 +1,41 @@
...
-
-    for (productID in productsLocal) {
-        let product = productsLocal[productID];
-        if(urlProductID == product.id) {
-            res.json(product)
-        }
-    }
-
-})
+    let urlProductID = req.params.id;
```
Comment (f4ww4z) [inline server.js]: Can't this be const? Also, use [object destructuring.](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment#Object_destructuring)

File: server.js [hunk 5]
```diff
@@ -1,49 +1,41 @@
...
 app.get('/category/:id', (req, res) => {
-    let urlCategoryID = req.params.id
-    let productsLocal = products.products
-    let arrayOfProducts = []
-
-    for (productID in productsLocal) {
-        let product = productsLocal[productID];
-        if(urlCategoryID == product.categoryId) {
-            arrayOfProducts.push(product)
+    let urlCategoryID = req.params.id;
```
Comment (f4ww4z) [inline server.js]: Can't this be const? Also, use [object destructuring.](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment#Object_destructuring)

File: server.js [hunk 6]
```diff
@@ -1,49 +1,41 @@
...
-    for (productID in productsLocal) {
-        let product = productsLocal[productID];
-        if(urlProductID == product.id) {
-            res.json(product)
-        }
-    }
-
-})
+    let urlProductID = req.params.id;
+    let productsMap = dataService.getCombinedProductMap();
```
Comment (f4ww4z) [inline server.js]: Can't it be const?

File: server.js [hunk 7]
```diff
@@ -1,49 +1,41 @@
...
-    let arrayOfProducts = []
-
-    for (productID in productsLocal) {
-        let product = productsLocal[productID];
-        if(urlCategoryID == product.categoryId) {
-            arrayOfProducts.push(product)
+    let urlCategoryID = req.params.id;
+    let combinedProductMap = dataService.getCombinedProductMap();
+    const productsBasedOnCategory = new Map();
+    combinedProductMap.forEach(((value, key, map) => {
```
Comment (f4ww4z) [inline server.js]: What is the extra `()` used for? What is `map` parameter for?

File: server.js [hunk 8]
```diff
@@ -1,49 +1,41 @@
...
+        if (value.categoryId === urlCategoryID) {
+            productsBasedOnCategory.set(key, value);
         }
-    }
+    }));
     
-    res.json(arrayOfProducts)
-})
+    res.send(productsBasedOnCategory);
```
Comment (f4ww4z) [inline server.js]: The output is in the wrong format. It should be

```json
{
  "P001": {
    "name": "...",
    "categoryId": "...",
    ...
  },
  "P002": {
    "name": "...",
    "categoryId": "...",
    ...
  }
}
```

File: server.js [hunk 9]
```diff
@@ -1,49 +1,41 @@
...
+    ];
     res.json(info)
-})
+});
 
 app.get('/products/all', (req, res) => {
-    res.json(products)
-})
+    let productsMap = dataService.getCombinedProductMap();
+    res.send(productsMap);
```
Comment (f4ww4z) [inline server.js]: Output should be in the format of

```json
{
  "P001": {
    "name": "...",
    "categoryId": "...",
    ...
  },
  "P002": {
    "name": "...",
    "categoryId": "...",
    ...
  }
}
```

File: server.js (hunk 8 above)
Comment (felixfavour) [reply server.js]: I told you it was rendered as `{ }` in my browser.

**Pull request closed without merging**